# 2. Bài tập thực hành

# Bài 1. Tạo Tensor cơ bản  
# Tạo một Tensor số nguyên kích thước 2 × 3 và hiển thị kết quả.

In [1]:
import torch

In [2]:
a= torch.tensor([[1, 2, 3], [4, 5, 6]],dtype=torch.int32)
print(a)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)


# Yêu cầu: Cho biết kích thước, số chiều và kiểu dữ liệu của Tensor a.

In [3]:
# Kích thước của Tensor a
print(a.size())
# Số chiều của Tensor a
print(a.dim())
# Kiểu dữ liệu của Tensor a
print(a.dtype)


torch.Size([2, 3])
2
torch.int32


# Bài 2. Tạo các Tensor thường dùng 
# Thực hành tạo Tensor ngẫu nhiên, Tensor toàn số 0, toàn số 1 và ma trận đơn vị.

In [4]:
import torch
a = torch.rand(3 , 4)
b= torch.randn(3 , 4)
d = torch.ones(2,3)
e = torch.eye(4)
print("Tensor ngẫu nhiên:\n", a)
print("Tensor chuẩn hóa:\n", b)
print("Tensor toàn số 1:\n", d)
print("Ma trận đơn vị:\n", e)

Tensor ngẫu nhiên:
 tensor([[0.4920, 0.2715, 0.1301, 0.5734],
        [0.5447, 0.3637, 0.6129, 0.2581],
        [0.1535, 0.3232, 0.6472, 0.3477]])
Tensor chuẩn hóa:
 tensor([[-0.1624,  1.3709,  1.0632, -1.0530],
        [-1.1600, -1.6013,  0.4067, -0.2704],
        [ 0.0568,  1.7812, -0.2362, -0.6189]])
Tensor toàn số 1:
 tensor([[1., 1., 1.],
        [1., 1., 1.]])
Ma trận đơn vị:
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])


# Yêu cầu:  
• Nhận xét sự khác nhau giữa torch.rand() và torch.randn().  
Sự khác nhau giữa torch.rand() và torch.randn() là:
- torch.rand() tạo ra các giá trị ngẫu nhiên từ phân phối đồng đều (uniform distribution) trong khoảng [0, 1).
- torch.randn() tạo ra các giá trị ngẫu nhiên từ phân phối chuẩn (normal distribution) với trung bình bằng 0 (mean = 0) và độ lệch chuẩn bằng 1 (std = 1).

# Tạo thêm một Tensor 5 × 5 chứa toàn giá trị 10.  

In [5]:
a = torch.full((5, 5), 10)
print("Tensor 5x5 chứa toàn giá trị 10:\n", a)

Tensor 5x5 chứa toàn giá trị 10:
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


# Bài 3. Kiểm tra thông tin Tensor  
# Sử dụng shape, size(), ndim và dtype để kiểm tra các thuộc tính của Tensor. 

In [6]:
import torch
a = torch.zeros(2,3,4)
# Kích thước của Tensor a
print(f"Kích thước của Tensor a: {a.shape}")
print(f"Kích thước của Tensor a (sử dụng size()): {a.size()}")
print(f"Số chiều của Tensor a: {a.ndim}")
print(f"Kiểu dữ liệu của Tensor a: {a.dtype}")


Kích thước của Tensor a: torch.Size([2, 3, 4])
Kích thước của Tensor a (sử dụng size()): torch.Size([2, 3, 4])
Số chiều của Tensor a: 3
Kiểu dữ liệu của Tensor a: torch.float32


# Yêu cầu: Giải thích ý nghĩa của torch.Size([2, 3, 4]).
torch.Size([2, 3, 4]) là một đối tượng đại diện cho kích thước của một Tensor. Nó cho biết Tensor có 3 chiều với các kích thước cụ thể như sau:
- Chiều đầu tiên có kích thước 2.
- Chiều thứ hai có kích thước 3.
- Chiều thứ ba có kích thước 4.
(Có thể hiểu là Tensor này chứa 2 "khối", mỗi khối có 3 hàng và 4 cột.)

# Bài 4. Slicing và Indexing  
# Truy xuất một phần dữ liệu từ Tensor nhiều chiều

In [7]:
import torch
a = torch.randn(3,4,5)
print("Original Tensor a:\n", a)
print("Tensor a[1]:\n", a[1])
print("\nTensor a[1:,2:4]:\n", a[1:,2:4])

Original Tensor a:
 tensor([[[-0.5174, -0.2004, -1.8617,  0.3604,  1.0793],
         [ 0.2741, -0.0155, -1.0284,  2.2082,  1.0816],
         [ 0.0052, -0.4206,  0.1941,  0.2899, -0.1430],
         [-0.5765, -0.1399, -0.4358,  0.5599, -1.2208]],

        [[-0.6493, -0.6713,  0.9831,  0.6327, -1.9112],
         [-0.9951, -0.7325, -0.9663, -1.3000, -1.3424],
         [ 0.7263, -1.8124, -0.6677, -0.2959, -0.8450],
         [-0.2073, -0.0405,  0.6727, -0.5244,  2.2220]],

        [[-0.5769, -0.2720,  0.8325, -0.2675, -0.7978],
         [-0.2347,  0.3952, -1.0073, -1.3943, -0.7378],
         [ 0.2393,  1.1477,  0.0753,  0.3095,  0.5277],
         [-0.8288,  1.1029,  0.1243, -0.4620,  0.7897]]])
Tensor a[1]:
 tensor([[-0.6493, -0.6713,  0.9831,  0.6327, -1.9112],
        [-0.9951, -0.7325, -0.9663, -1.3000, -1.3424],
        [ 0.7263, -1.8124, -0.6677, -0.2959, -0.8450],
        [-0.2073, -0.0405,  0.6727, -0.5244,  2.2220]])

Tensor a[1:,2:4]:
 tensor([[[ 0.7263, -1.8124, -0.6677, -0.2959, -

# Yêu cầu:  
• Cho biết shape của a[1] và a[1:, 2:4].  
• Thử lấy hai phần tử cuối cùng trên chiều thứ ba.  

In [8]:
# shape của a[1] và a[1:, 2:4]
print(f"Shape của a[1]: {a[1].shape}")
print(f"Shape của a[1:, 2:4]: {a[1:, 2:4].shape}")

# Lấy hai phần tử cuối cùng trên chiều thứ ba
print(f"Shape của a[:, :, -2:]: {a[:, :, -2:].shape}")

Shape của a[1]: torch.Size([4, 5])
Shape của a[1:, 2:4]: torch.Size([2, 2, 5])
Shape của a[:, :, -2:]: torch.Size([3, 4, 2])


# Bài 5. Reshape và Flatten  
Thay đổi hình dạng Tensor mà không làm thay đổi số phần tử.

In [9]:
import torch
a = torch.randn(3,4)
print("Original Tensor a:\n", a)
# ravel() trả về một Tensor 1 chiều chứa 
# tất cả các phần tử của Tensor a
print("Flatten: ", a.ravel())
print("\nReshape to 3 x 2 x 2: ", a.reshape(3, 2, 2))


Original Tensor a:
 tensor([[-0.2415,  0.6972, -0.7306, -1.0248],
        [ 0.5888, -1.4030, -1.0502, -0.4338],
        [ 1.2685, -0.7444,  0.9900,  1.3113]])
Flatten:  tensor([-0.2415,  0.6972, -0.7306, -1.0248,  0.5888, -1.4030, -1.0502, -0.4338,
         1.2685, -0.7444,  0.9900,  1.3113])

Reshape to 3 x 2 x 2:  tensor([[[-0.2415,  0.6972],
         [-0.7306, -1.0248]],

        [[ 0.5888, -1.4030],
         [-1.0502, -0.4338]],

        [[ 1.2685, -0.7444],
         [ 0.9900,  1.3113]]])


# Yêu cầu:  
• Kiểm tra số phần tử trước và sau khi reshape.  
• Thử reshape Tensor thành kích thước 2 × 6. 

In [10]:
x = torch.arange(24)
print("Original Tensor x:\n", x)
print(f"Số phần tử của x: {x.numel()}")
print("\nReshape to 2 x 3 x 4: ", x.reshape(2, 3, 4))
print(f"Số phần tử của x sau khi reshape: {x.reshape(2, 3, 4).numel()}")
print(f"\nReshape thành 2 x 6: {a.reshape(2, 6)}")

Original Tensor x:
 tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])
Số phần tử của x: 24

Reshape to 2 x 3 x 4:  tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])
Số phần tử của x sau khi reshape: 24

Reshape thành 2 x 6: tensor([[-0.2415,  0.6972, -0.7306, -1.0248,  0.5888, -1.4030],
        [-1.0502, -0.4338,  1.2685, -0.7444,  0.9900,  1.3113]])


# Bài 6. Phép toán và nhân ma trận  
Thực hiện các phép toán phần tử và phép nhân ma trận bằng toán tử @.

In [11]:
import  torch  
a  =  torch.randn(2,  3)  
b  = torch.randn(2, 3) 
print("a: \n", a)
print("b: \n", b)

a: 
 tensor([[-1.2460, -1.3097, -1.2027],
        [ 0.7582,  0.0230,  0.5112]])
b: 
 tensor([[-0.0829,  1.3476,  0.2445],
        [-0.0838, -1.6085,  1.2098]])


In [12]:
print("\na + b:")  
print(a + b)  
print("\na / b:",a / b) 
print("\na^2:",a  **  2)   
print("\nMatrix  multiplication:") 
print(a @ b.T)


a + b:
tensor([[-1.3289,  0.0378, -0.9581],
        [ 0.6744, -1.5855,  1.7210]])

a / b: tensor([[ 1.5039e+01, -9.7192e-01, -4.9184e+00],
        [-9.0450e+00, -1.4295e-02,  4.2257e-01]])

a^2: tensor([[1.5526e+00, 1.7154e+00, 1.4464e+00],
        [5.7488e-01, 5.2868e-04, 2.6135e-01]])

Matrix  multiplication:
tensor([[-1.9558,  0.7562],
        [ 0.0932,  0.5179]])


# Yêu cầu:  
• Phân biệt phép toán a * b với a @ b.T
   - a * b là phép nhân phần tử (element-wise multiplication), nghĩa là mỗi phần tử của Tensor a sẽ được nhân với phần tử tương ứng của Tensor b. Kết quả sẽ có cùng kích thước với a và b.
   - a @ b.T là phép nhân ma trận (matrix multiplication), nghĩa là Tensor a sẽ được nhân với chuyển vị của Tensor b. Kết quả sẽ có kích thước khác, phụ thuộc vào số hàng của a và số cột của b.T
   
• Cho biết shape của kết quả a @ b.T. 

In [13]:
y = a * b 
print("Kết quả a * b:\n", y)
y = a @ b.T
print("Kết quả a @ b.T:\n", y)
print(f"Shape của kết quả a @ b.T: {y.shape}")

Kết quả a * b:
 tensor([[ 0.1032, -1.7650, -0.2941],
        [-0.0636, -0.0370,  0.6185]])
Kết quả a @ b.T:
 tensor([[-1.9558,  0.7562],
        [ 0.0932,  0.5179]])
Shape của kết quả a @ b.T: torch.Size([2, 2])


# Bài 7. Các hàm thống kê trên Tensor  
Tính trung bình, độ lệch chuẩn và tổng tích lũy theo từng cột.  

In [14]:
import torch
a = torch.randn(3,4)
print(a)
print("\nMean: ",torch.mean(a,dim=0))  # Tính trung bình theo từng cột
print("\nStandard Deviation: ",torch.std(a,dim=0))  # Tính độ lệch chuẩn theo từng cột
print("\nCumulative Sum: ", torch.cumsum(a, dim=0))  # Tổng tích lũy theo chiều dọc

tensor([[ 0.8285, -1.4244, -0.3989, -0.4181],
        [-0.7344, -0.9755, -0.4329,  0.3754],
        [-0.8023, -0.8070,  1.9072, -0.4118]])

Mean:  tensor([-0.2361, -1.0689,  0.3585, -0.1515])

Standard Deviation:  tensor([0.9226, 0.3192, 1.3414, 0.4563])

Cumulative Sum:  tensor([[ 0.8285, -1.4244, -0.3989, -0.4181],
        [ 0.0941, -2.3999, -0.8318, -0.0427],
        [-0.7083, -3.2068,  1.0755, -0.4545]])


# Yêu cầu:  
• Thay dim=0 bằng dim=1 và nhận xét sự khác nhau. 

In [15]:
import torch
a = torch.randn(3,4)
print(a)
print("\nMean with dim=0: ",torch.mean(a,dim=0))  # Tính trung bình theo từng cột
print("\nMean with dim=1: ",torch.mean(a,dim=1))  # Tính trung bình theo từng hàng
print("\nStandard Deviation with dim=0: ",torch.std(a,dim=0))  # Tính độ lệch chuẩn theo từng cột
print("\nStandard Deviation with dim=1: ",torch.std(a,dim=1))  # Tính độ lệch chuẩn theo từng hàng
print("\nCumulative Sum with dim=0: ", torch.cumsum(a, dim=0))  # Tích lũy theo chiều dọc
print("\nCumulative Sum with dim=1: ", torch.cumsum(a, dim=1))  # Tích lũy theo chiều ngang

tensor([[ 1.3539, -0.6524, -0.8374,  1.3343],
        [-1.1034, -0.0428, -0.8127, -0.6379],
        [-0.9730, -1.4973, -0.7647,  0.1951]])

Mean with dim=0:  tensor([-0.2408, -0.7308, -0.8050,  0.2972])

Mean with dim=1:  tensor([ 0.2996, -0.6492, -0.7600])

Standard Deviation with dim=0:  tensor([1.3826, 0.7304, 0.0370, 0.9900])

Standard Deviation with dim=1:  tensor([1.2085, 0.4475, 0.7074])

Cumulative Sum with dim=0:  tensor([[ 1.3539, -0.6524, -0.8374,  1.3343],
        [ 0.2506, -0.6952, -1.6502,  0.6964],
        [-0.7225, -2.1925, -2.4149,  0.8915]])

Cumulative Sum with dim=1:  tensor([[ 1.3539,  0.7016, -0.1359,  1.1984],
        [-1.1034, -1.1462, -1.9589, -2.5968],
        [-0.9730, -2.4703, -3.2351, -3.0400]])


## Nhận xét

- Với `dim=0`, `mean()` và `std()` gộp các hàng để tính theo từng cột nên kết quả có shape `[4]`. `cumsum()` cộng dồn từ trên xuống dưới và vẫn giữ shape `[3, 4]`.
- Với `dim=1`, `mean()` và `std()` gộp các cột để tính theo từng hàng nên kết quả có shape `[3]`. `cumsum()` cộng dồn từ trái sang phải và vẫn giữ shape `[3, 4]`.

# Bài 8. Autograd cơ bản  
Tạo biến có requires_grad=True và dùng backward() để tính đạo hàm tự động.

In [16]:
import torch
x = torch.tensor(3.6, requires_grad=True)
y= x * x
y.backward()  # Tính đạo hàm tự động
print("x = ", x)
print("y = ", y)
print("x.grad = ", x.grad)  # In ra đạo hàm của y theo x

x =  tensor(3.6000, requires_grad=True)
y =  tensor(12.9600, grad_fn=<MulBackward0>)
x.grad =  tensor(7.2000)


C:\ProgramData\Anaconda3\Lib\site-packages\torch\autograd\graph.py:1072: UserWarning: CUDA initialization: The NVIDIA driver on your system is too old (found version 11060). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\c10\cuda\CUDAFunctions.cpp:119.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass


# Yêu cầu:  
• Tính thủ công đạo hàm của y = x² tại x = 3.6 và so sánh với x.grad
y = x²
y' = 2x 
Với x = 3.6, ta có y' = 2 * 3.6 = 7.2
So với x.grad = 7.2000 => Kết quả tính đạo hàm tự động bằng backward() là chính xác so với đạo hàm tính thủ công.
• Đổi hàm thành y = 3*x**2 + 2*x + 1 và kiểm tra gradient. 

In [17]:
import torch

x = torch.tensor(3.6, requires_grad=True)
y = 3 * x ** 2 + 2 * x + 1
y.backward()
print("x = ", x)
print("y = ", y)
print("x.grad = ", x.grad)

x =  tensor(3.6000, requires_grad=True)
y =  tensor(47.0800, grad_fn=<AddBackward0>)
x.grad =  tensor(23.6000)


# Bài 9. Gradient Descent tìm hệ số đa thức 
Sinh dữ liệu từ đa thức y = x² + 2x + 3, sau đó dùng NAdam và autograd để tìm lại ba hệ số. 

In [18]:
import numpy as np 
import torch
polynomial = np.poly1d([1, 2, 3])  # Tạo đa thức y = x² + 2x + 3
N = 20
x = np.random.randn(N,1) * 5 
y = polynomial(x)
XX =np.hstack([x*x, x, np.ones_like(x)])
w = torch.randn(3,1,requires_grad=True)
x= torch.tensor(XX, dtype=torch.float32)
y= torch.tensor(y, dtype=torch.float32)
optimizer = torch.optim.NAdam([w], lr=0.01)
print("Initial Coefficients: ", w)
for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w  
    mse = torch.mean(torch.square(y_pred - y))
    mse.backward()
    optimizer.step()
    if epoch % 100 == 0:
        print(f"Epoch {epoch}, MSE: {mse.item():.6f}, w: {w.detach().flatten()}")

Initial Coefficients:  tensor([[ 0.7268],
        [ 0.1142],
        [-0.7365]], requires_grad=True)
Epoch 0, MSE: 122.749390, w: tensor([ 0.7373,  0.1247, -0.7259])
Epoch 100, MSE: 20.194565, w: tensor([ 1.0453,  1.0098, -0.0830])
Epoch 200, MSE: 6.536692, w: tensor([1.0463, 1.5824, 0.5140])
Epoch 300, MSE: 2.710241, w: tensor([1.0403, 1.8575, 1.0873])
Epoch 400, MSE: 1.354810, w: tensor([1.0314, 1.9583, 1.5880])


Epoch 500, MSE: 0.673405, w: tensor([1.0227, 1.9873, 1.9990])
Epoch 600, MSE: 0.311375, w: tensor([1.0155, 1.9948, 2.3193])
Epoch 700, MSE: 0.132302, w: tensor([1.0101, 1.9971, 2.5564])
Epoch 800, MSE: 0.051470, w: tensor([1.0063, 1.9983, 2.7235])
Epoch 900, MSE: 0.018284, w: tensor([1.0038, 1.9990, 2.8353])


# Yêu cầu:  
• So sánh ba hệ số học được với [1, 2, 3]
    Ba hệ số học được gần với [1, 2, 3] cụ thể với kết quả ở lần 900 với w :  w: tensor([1.0008, 2.0001, 2.9439]), điều này chứng tỏ rằng quá trình huấn luyện bằng NAdam và autograd đã thành công trong việc tìm lại các hệ số của đa thức gốc.
• Giải thích vai trò của optimizer.zero_grad(), mse.backward() và optimizer.step()
    optimizer.zero_grad(): Xóa các gradient đã tính toán trước đó để tránh việc cộng dồn gradient từ các bước huấn luyện trước.
    mse.backward(): Tính toán gradient của hàm mất mát (mean squared error) đối với các tham số của mô hình.
    optimizer.step(): Cập nhật các tham số của mô hình dựa trên gradient đã tính toán và thuật toán tối ưu hóa (NAdam trong trường hợp này).

# Bài 10. Giải hệ phương trình bằng Autograd  
Tối thiểu hóa tổng sai số bình phương để tìm A, B, C, D thỏa bốn phương trình.
    Kiểm tra nghiệm thu được với bốn phương trình: A+B=9, C-D=1,   A+C=8, B-D=2

In [19]:
import random
import torch
A = torch.tensor(random.random(),requires_grad=True)
B = torch.tensor(random.random(),requires_grad=True)
C = torch.tensor(random.random(),requires_grad=True)
D = torch.tensor(random.random(),requires_grad=True)
EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)  
for epoch in range(EPOCHS):   
    y1 = A + B - 9   
    y2 = C - D - 1 
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    optimizer.zero_grad()  
    sqerr.backward()  
    optimizer.step()  
    if epoch % 100 == 0:
        print(f"Epoch {epoch}, sqerr = {sqerr.item():.8f}")

# Kiểm tra nghiệm sau khi tối ưu
with torch.no_grad():
    errors = torch.stack([
        A + B - 9,
        C - D - 1,
        A + C - 8,
        B - D - 2
    ])

    print("\nNghiệm tìm được:")
    print(f"A = {A.item():.4f}")
    print(f"B = {B.item():.4f}")
    print(f"C = {C.item():.4f}")
    print(f"D = {D.item():.4f}")
    print("Các sai số:", errors)
    print("Tổng sai số bình phương:", torch.sum(errors**2).item())

Epoch 0, sqerr = 102.09529114
Epoch 100, sqerr = 55.06598282
Epoch 200, sqerr = 28.55310822


Epoch 300, sqerr = 13.42156029
Epoch 400, sqerr = 5.65199852
Epoch 500, sqerr = 2.12208223
Epoch 600, sqerr = 0.70800018


Epoch 700, sqerr = 0.20943613
Epoch 800, sqerr = 0.05482016
Epoch 900, sqerr = 0.01266345
Epoch 1000, sqerr = 0.00257294


Epoch 1100, sqerr = 0.00045772
Epoch 1200, sqerr = 0.00007086
Epoch 1300, sqerr = 0.00000950
Epoch 1400, sqerr = 0.00000109


Epoch 1500, sqerr = 0.00000011
Epoch 1600, sqerr = 0.00000001
Epoch 1700, sqerr = 0.00000000


Epoch 1800, sqerr = 0.00000000
Epoch 1900, sqerr = 0.00000000

Nghiệm tìm được:
A = 4.3539
B = 4.6460
C = 3.6460
D = 2.6460
Các sai số: tensor([-4.5776e-05,  1.4305e-05, -2.2411e-05, -9.5367e-06])
Tổng sai số bình phương: 2.893330020015128e-09


#  Nhận xét vì sao bài toán có thể có nhiều nghiệm
Sau quá trình tối ưu, tổng sai số bình phương tiến gần về 0 và các giá trị A, B, C, D tìm được thỏa bốn phương trình.

Tuy nhiên, hệ không có nghiệm duy nhất vì phương trình thứ tư phụ thuộc vào ba phương trình còn lại. Nghiệm tổng quát của hệ là:

(A, B, C, D) = (t, 9-t, 8-t, 7-t)

Do đó, mỗi lần chạy chương trình có thể thu được một bộ giá trị khác nhau do khởi tạo ngẫu nhiên, nhưng chúng vẫn thỏa hệ nếu tổng sai số bình phương gần bằng 0.

# Bài 11. Viết chương trình PyTorch để thực hiện các yêu cầu sau:  
• Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 × 6, 2 × 12 và 2 × 3 × 4.  
• Tạo hai Tensor A, B kích thước 3 × 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột. 

 

In [20]:
import torch
a = torch.arange(24)
print("Original Tensor a:\n", a)
print("reshape to 4 x 6:\n", a.reshape(4, 6))
print("reshape to 2 x 12:\n", a.reshape(2, 12))
print("reshape to 2 x 3 x 4:\n", a.reshape(2, 3, 4))
A = torch.randn(3,4)
B = torch.randn(3,4)
print("Tinh A + B:\n", A + B)
print("Tinh A * B:\n", A * B)
print("Tinh A / B:\n", A / B)
print("Trung binh theo cot:\n", A.mean(dim=0))
print("Do lech chuan theo cot:\n", A.std(dim=0))



Original Tensor a:
 tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])
reshape to 4 x 6:
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])
reshape to 2 x 12:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])
reshape to 2 x 3 x 4:
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])
Tinh A + B:
 tensor([[ 0.3238, -0.8661,  0.5894, -0.6862],
        [ 0.6803, -0.8624,  1.1830, -0.9816],
        [-0.0468, -2.1892, -0.6292, -0.1844]])
Tinh A * B:
 tensor([[-1.0988e+00, -1.8688e+00,  1.6758e-03,  9.0086e-02],
        [ 1.1454e-01, -3.3244e+00,  3.4117e-01, -2.0938e-01],
        [-2.0028e-01,  1.1029e+00, -6.9244e-01, -3.8583e+00]])
Tinh A / B:
 tensor([[-0.7351, -1

In [21]:
# Cho y = 2x² + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và
#  kiểm tra bằng phép tính thủ công. 
import torch
x = torch.tensor(2.0, requires_grad=True)
y = 2 * x ** 2 + 5 * x + 3
y.backward()
print("dy/dx at x=2:", x.grad.item())

dy/dx at x=2: 13.0


### Kiểm tra đạo hàm bằng phép tính thủ công

Với $y=2x^2+5x+3$, ta có:

$$
\frac{dy}{dx}=4x+5.
$$

Tại $x=2$:

$$
\frac{dy}{dx}=4(2)+5=13.
$$

Kết quả `x.grad = 13` từ Autograd trùng với kết quả tính thủ công.

In [22]:
# Thay đa thức trong Bài 9 bằng y = 2x² - 3x + 5; 
# huấn luyện và cho biết các hệ số tìm được. 
import numpy as np
import torch

np.random.seed(42)
torch.manual_seed(42)

polynomial = np.poly1d([2, -3, 5])  # y = 2x² - 3x + 5
N = 20
x_data = np.random.randn(N, 1) * 5
y_data = polynomial(x_data)
X_data = np.hstack([x_data**2, x_data, np.ones_like(x_data)])

X = torch.tensor(X_data, dtype=torch.float32)
y = torch.tensor(y_data, dtype=torch.float32)
w = torch.randn(3, 1, requires_grad=True)

optimizer = torch.optim.NAdam([w], lr=0.01)
EPOCHS = 5000

print("Hệ số ban đầu:", w.detach().flatten())
for epoch in range(EPOCHS):
    optimizer.zero_grad()
    y_pred = X @ w
    mse = torch.mean((y_pred - y) ** 2)
    mse.backward()
    optimizer.step()

    if epoch % 500 == 0:
        print(f"Epoch {epoch:4d}, MSE: {mse.item():.6f}, w: {w.detach().flatten()}")

learned = w.detach().flatten()
expected = torch.tensor([2.0, -3.0, 5.0])
print("\nHệ số tìm được:", learned)
print("Hệ số mong đợi:", expected)
print("Sai lệch tuyệt đối:", torch.abs(learned - expected))

Hệ số ban đầu: tensor([0.3367, 0.1288, 0.2345])
Epoch    0, MSE: 4697.240723, w: tensor([0.3473, 0.1182, 0.2450])


Epoch  500, MSE: 13.002596, w: tensor([ 2.0787, -2.3414,  2.2242])


Epoch 1000, MSE: 3.123401, w: tensor([ 2.0462, -2.8946,  2.7765])
Epoch 1500, MSE: 1.474474, w: tensor([ 2.0296, -2.9830,  3.4279])


Epoch 2000, MSE: 0.562867, w: tensor([ 2.0181, -2.9922,  4.0286])
Epoch 2500, MSE: 0.155707, w: tensor([ 2.0095, -2.9959,  4.4893])


Epoch 3000, MSE: 0.028729, w: tensor([ 2.0041, -2.9983,  4.7807])
Epoch 3500, MSE: 0.003164, w: tensor([ 2.0014, -2.9994,  4.9273])


Epoch 4000, MSE: 0.000180, w: tensor([ 2.0003, -2.9999,  4.9827])
Epoch 4500, MSE: 0.000004, w: tensor([ 2.0001, -3.0000,  4.9973])



Hệ số tìm được: tensor([ 2.0000, -3.0000,  4.9997])
Hệ số mong đợi: tensor([ 2., -3.,  5.])
Sai lệch tuyệt đối: tensor([5.4836e-06, 6.9141e-06, 2.6369e-04])


## Tính thủ công các hệ số

Giả sử chưa biết các hệ số của đa thức và đặt:

$$
y=ax^2+bx+c.
$$

Chọn ba giá trị đơn giản $x=0$, $x=1$ và $x=2$ từ đa thức $y=2x^2-3x+5$:

- Với $x=0$: $y=5$, do đó $c=5$.
- Với $x=1$: $y=4$, do đó $a+b+c=4 \Rightarrow a+b=-1$.
- Với $x=2$: $y=7$, do đó $4a+2b+c=7 \Rightarrow 2a+b=1$.

Lấy $(2a+b)-(a+b)=1-(-1)$, ta được $a=2$. Thay vào $a+b=-1$:

$$
b=-1-2=-3.
$$

Vậy ba hệ số tính thủ công là:

$$
\boxed{[a,b,c]=[2,-3,5]}.
$$

Sau khi huấn luyện, các hệ số trong `w` phải xấp xỉ `[2, -3, 5]`. Chúng có thể sai lệch rất nhỏ vì NAdam tìm nghiệm bằng cách cập nhật lặp, thay vì giải hệ phương trình trực tiếp.